File: 938-range-sum-of-bst/938-range-sum-of-bst.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void preOrder(struct TreeNode* root, int low, int high, int *sum){
    if(root == NULL) return;
    if((root->val >= low) && (root->val <= high))
        *sum += root->val;
    if(root->left != NULL)
        preOrder(root->left,low,high,sum);
    if(root->right != NULL)
        preOrder(root->right,low,high,sum);
}

int rangeSumBST(struct TreeNode* root, int low, int high){
    int sum = 0;
    preOrder(root,low,high,&sum);
    return sum;
}

File: 938-range-sum-of-bst/938-range-sum-of-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        //Integer sum = 0;  -->immutable object
        //int sum = 0;      -->call by reference DOES'NT work in Java
        MyVariable obj = new MyVariable();
        preorder_traverse(root, low, high, obj);   
        return obj.sum;
    }
    
    public void preorder_traverse(TreeNode root, int low, int high, MyVariable obj) {
        if (root == null)
            return ;
        if (root.val >= low && root.val <= high)
            obj.sum += root.val;
        preorder_traverse(root.left, low, high, obj);
        preorder_traverse(root.right, low, high, obj);
    }
}

class MyVariable {
    int sum;  //default value of int instance variable: 0
}